[16.01] libraries: handle correctly application/json pased in POST

martenson opened the pull request:
This is an attempt to properly fix: https://github.com/galaxyproject/galaxy/pull/1629

ping @gregvonkuster @blankenberg

martenson committed:
handle correctly application/json pased in POST

jmchilton commented:
I preferred Greg's documentation changes - it seemed more inline with most of the rest of the API. Still this is a good change on its own.

:+1:

martenson commented:
You can create a POST request with url parameters, then the payload is non-existent, but the API should still work imo.

jmchilton commented:
For nearly all other Galaxy POST requests - we consume request properties via payload and ignore URL parameters. If it needs to work for this endpoint for backward compatibility that is totally fine - but we shouldn't encourage it IMO.

martenson commented:
@gregvonkuster please test this PR with your implementation thanjks

gregvonkuster commented:
@martenson This fix worked for my post.  You may need to make similar changes in other POST methods for libraries.  Thanks Martin!  ;)

martenson committed:
fix setting library permissions to accept JSON correctly

martenson committed:
unify api for library creation; support POST with url params

martenson committed:
catch value errors from id decoding

martenson committed:
fix updating library API to not require payload
and to properly handle optional params

martenson committed:
allow specifying DELETE params in JSON

martenson committed:
allow creating folders with JSON payload

martenson committed:
accept json as input for setting permissions;
fix inconsistency by leveraging role_manager

martenson committed:
allow PATCHing with payload in urlparams

martenson commented:
updated and rebased

I tried to fix all library API endpoints accepting POST, PUT, PATCH, DELETE where it made sense to accept json

I also fixed a remnant of poor Role handling in folders API controller, it now uses Role manager similar to libraries controller

martenson committed:
remove deletion of the payload dict item
it is not necessary and the key does not have to exist

martenson committed:
fix unused imports and unused vars

nsoranzo commented on lib/galaxy/webapps/galaxy/api/libraries.py:
```diff
@@ -239,7 +240,7 @@ def get_permissions( self, trans, encoded_library_id, **kwd ):
-    def set_permissions( self, trans, encoded_library_id, **kwd ):
+    def set_permissions( self, trans, encoded_library_id, payload=None, **kwd ):
         """
         def set_permissions( self, trans, encoded_dataset_id, **kwd ):
```
This doc line is not in sync with the method signature, I'd suggest to remove it.

martenson replied:
will fix

martenson committed:
delete the 'payload' if it is in kwd so it is not supplied twice

martenson committed:
Revert "delete the 'payload' if it is in kwd so it is not supplied twice"

This reverts commit 134d792c1fcfba5d09e2937074c3263792bea991.

martenson committed:
update docs of the data libraries API

martenson commented:
updated documentation of all touched methods

martenson commented:
please review and merge, this needs to go to release

martenson commented:
thanks John
